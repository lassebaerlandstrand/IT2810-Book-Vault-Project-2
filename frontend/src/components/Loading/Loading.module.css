:where([data-mantine-color-scheme='light']) .icon {
  filter: invert(1);
}
