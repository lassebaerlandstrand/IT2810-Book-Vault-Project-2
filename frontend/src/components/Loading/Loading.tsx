import { forwardRef } from 'react';
import { Image, Loader, LoaderProps, MantineLoaderComponent, Stack, Text } from '@mantine/core';
import BookLoaderGif from '../../assets/BookLoader.gif';
import styles from './Loading.module.css';

/** Define a new Mantine Loader. This will be used in theme.ts, such that the Loading component below has access to this and we follow Mantine best practices */
export const BookLoader: MantineLoaderComponent = forwardRef(({ style, ...others }, ref) => (
  <Image
    src={BookLoaderGif}
    {...others}
    className={`${styles.icon} ${others.className}`}
    ref={ref}
    style={{
      width: 'var(--loader-size)',
      height: 'var(--loader-size)',
      stroke: 'var(--loader-color)',
      ...style,
    }}
  />
));

/**
 * Loading component which has custom styling which you can use directly.
 * Displays a flipping book as a loader.
 */
const Loading = (props: LoaderProps) => {
  return (
    <Stack my={100} justify="center" align="center">
      <Loader size="xxl" {...props} type="book" />
      <Text fw={500}>Loading</Text>
    </Stack>
  );
};

export default Loading;
