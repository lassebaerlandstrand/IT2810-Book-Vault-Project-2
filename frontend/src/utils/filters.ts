import { SortBy, SortOrder } from '@/generated/graphql';

export const DEFAULT_FILTERS = {
  sortBy: SortBy.BookName,
  sortOrder: SortOrder.Asc,
  selectedGenres: [],
  selectedPublishers: [],
  selectedAuthors: [],
  selectedBeforeDate: new Date(2020, 11, 31),
  selectedAfterDate: new Date(1900, 0, 1),
  selectedMinPages: undefined,
  selectedMaxPages: undefined,
  selectedMinRating: undefined,
};

export const getFilterParams = (searchParams: URLSearchParams) => {
  const sortBy = (searchParams.get('sortBy') as SortBy) ?? DEFAULT_FILTERS.sortBy;
  const sortOrder = (searchParams.get('sortOrder') as SortOrder) ?? DEFAULT_FILTERS.sortOrder;
  const authors = searchParams.getAll('authors') ?? DEFAULT_FILTERS.selectedAuthors;
  const publishers = searchParams.getAll('publishers') ?? DEFAULT_FILTERS.selectedPublishers;
  const genres = searchParams.getAll('genres') ?? DEFAULT_FILTERS.selectedGenres;
  const beforeDate = searchParams.get('beforeDate')
    ? new Date(searchParams.get('beforeDate') ?? '')
    : DEFAULT_FILTERS.selectedBeforeDate;
  const afterDate = searchParams.get('afterDate')
    ? new Date(searchParams.get('afterDate') ?? '')
    : DEFAULT_FILTERS.selectedAfterDate;
  const minPages =
    parseInt(searchParams.get('minPages') ?? '', 10) || DEFAULT_FILTERS.selectedMinPages;
  const maxPages =
    parseInt(searchParams.get('maxPages') ?? '', 10) || DEFAULT_FILTERS.selectedMaxPages;
  const minRating =
    parseInt(searchParams.get('minRating') ?? '', 10) || DEFAULT_FILTERS.selectedMinRating;

  return {
    sortBy,
    sortOrder,
    authors,
    publishers,
    genres,
    beforeDate,
    afterDate,
    minPages,
    maxPages,
    minRating,
  };
};
