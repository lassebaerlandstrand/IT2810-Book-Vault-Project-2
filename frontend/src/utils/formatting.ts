/**
 * Formats a number by adding spaces as thousand separators.
 *
 * @param {number} number - The input number to format
 * @returns {string} The formatted number string with space separators
 *
 * @example
 * // returns '1 000'
 * formatNumberWithSpaces(1000);
 *
 * @example
 * // returns '1 000 000'
 * formatNumberWithSpaces(1000000);
 */
export const formatNumberWithSpaces = (number: number) =>
  number.toString().replace(/\B(?=(\d{3})+(?!\d))/g, ' ');

// The regular expression used in the replace method:
// \B - matches a non-word boundary to avoid adding spaces at the start of the string.
// (?=(\d{3})+(?!\d)) - positive lookahead to find positions followed by a group of three digits not followed by another digit.
