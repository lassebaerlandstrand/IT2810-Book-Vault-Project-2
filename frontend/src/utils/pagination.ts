const DEFAULT_PAGE = 1;
const DEFAULT_LIMIT = 25;
const LIMIT_OPTIONS = ['10', '25', '50', '100'];

export const getPaginationParams = (searchParams: URLSearchParams) => {
  const page = Number(searchParams.get('page')) || DEFAULT_PAGE;
  const limit = Number(searchParams.get('limit')) || DEFAULT_LIMIT;

  return { page, limit, DEFAULT_PAGE, DEFAULT_LIMIT, LIMIT_OPTIONS };
};
