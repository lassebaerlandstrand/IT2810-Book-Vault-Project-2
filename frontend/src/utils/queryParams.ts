import { SetURLSearchParams } from 'react-router-dom';

/**
 * Updates the URL search parameters based on the provided key and value.
 * Handles different types of values (single string, array of strings, or undefined).
 * This function handles consecutive calls.
 *
 * @param {SetURLSearchParams} setSearchParams - Function to update the URL search parameters
 * @param {string} key - The key of the query parameter to update
 * @param {string | string[] | undefined} value - The value to set for the query parameter
 *
 * @example
 * updateQueryParams(setSearchParams, 'search', 'newSearchQuery')
 * updateQueryParams(setSearchParams, 'genres', ['tag1', 'tag2'])
 * updateQueryParams(setSearchParams, 'page', '5')
 * updateQueryParams(setSearchParams, 'page', undefined) // Removes the key
 *
 */
export const updateQueryParams = (
  setSearchParams: SetURLSearchParams,
  key: string,
  value: string | string[] | undefined
) => {
  setSearchParams((params) => {
    if (Array.isArray(value)) {
      params.delete(key);
      value.forEach((v) => params.append(key, v));
    } else if (value === undefined || value === '') {
      params.delete(key);
    } else {
      params.set(key, value);
    }
    return params;
  });
};
