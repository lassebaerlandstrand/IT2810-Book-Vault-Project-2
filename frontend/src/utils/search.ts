/**
 * Extracts the search parameter from the URL.
 * This function ensures that the search parameters are correctly parsed and returned in a consistent format.
 *
 * @param {URLSearchParams} searchParams - URL search parameters containing search criteria
 * @returns {Object} Search parameters
 *
 */
export const getSearchParams = (searchParams: URLSearchParams) => {
  const searchValue = searchParams.get('search') || '';
  return { searchValue };
};
