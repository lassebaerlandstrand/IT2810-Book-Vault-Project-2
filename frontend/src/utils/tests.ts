/**
 * Removes random attributes generated by Mantine from DOM elements.
 * This is necessary to ensure consistent snapshots in tests, as Mantine uses random IDs.
 */
export const removeMantineRandomAttributes = () => {
  const attributesToRemove = [
    ...document.body.querySelectorAll('div [id^="mantine"]'),
    ...document.body.querySelectorAll('div [for^="mantine"]'),
  ];
  attributesToRemove.forEach((element) => {
    element.removeAttribute('for');
    element.removeAttribute('id');
    element.removeAttribute('class');
    element.removeAttribute('aria-describedby');
  });
};
