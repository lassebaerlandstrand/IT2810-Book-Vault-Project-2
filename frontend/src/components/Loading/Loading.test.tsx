import { render } from '@test-utils';
import { describe, expect, it } from 'vitest';
import { removeMantineRandomAttributes } from '@/utils/tests';
import Loading from './Loading';

describe('Loading', () => {
  it('renders a centered loader with the correct props', () => {
    const container = render(<Loading />);

    const loaderElement = container.baseElement.getElementsByClassName('mantine-Loader-root')[0];
    expect(loaderElement).toBeInTheDocument();
  });

  it('matches snapshot', () => {
    const { asFragment } = render(<Loading />);
    removeMantineRandomAttributes();
    expect(asFragment()).toMatchSnapshot();
  });
});
