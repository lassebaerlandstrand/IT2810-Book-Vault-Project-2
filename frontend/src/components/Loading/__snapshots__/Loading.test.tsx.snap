// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Loading > matches snapshot 1`] = `
<DocumentFragment>
  <style
    data-mantine-styles="true"
  >
    :root{--mantine-font-family: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-font-family-headings: Poppins, Roboto, Helvetica, Arial, sans-serif, Apple Color Emoji, Segoe UI Emoji;--mantine-breakpoint-xs: 30em;--mantine-breakpoint-md: 64em;--mantine-breakpoint-xl: 90em;--mantine-breakpoint-xxs: 24em;--mantine-color-logo-theme-0: #fefefe;--mantine-color-logo-theme-1: #cbcfd2;--mantine-color-logo-theme-2: #97a0a6;--mantine-color-logo-theme-3: #7d8991;--mantine-color-logo-theme-4: #63717b;--mantine-color-logo-theme-5: #495a65;--mantine-color-logo-theme-6: #2f434f;--mantine-color-logo-theme-7: #132937;--mantine-color-logo-theme-8: #0b202e;--mantine-color-logo-theme-9: #061017;}:root[data-mantine-color-scheme="dark"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-8);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-9);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.15);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.2);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-3);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-4);--mantine-color-logo-theme-outline-hover: rgba(99, 113, 123, 0.05);}:root[data-mantine-color-scheme="light"]{--mantine-color-logo-theme-text: var(--mantine-color-logo-theme-filled);--mantine-color-logo-theme-filled: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-filled-hover: var(--mantine-color-logo-theme-7);--mantine-color-logo-theme-light: rgba(47, 67, 79, 0.1);--mantine-color-logo-theme-light-hover: rgba(47, 67, 79, 0.12);--mantine-color-logo-theme-light-color: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline: var(--mantine-color-logo-theme-6);--mantine-color-logo-theme-outline-hover: rgba(47, 67, 79, 0.05);}
  </style>
  <style
    data-mantine-styles="classes"
  >
    @media (max-width: 29.99375em) {.mantine-visible-from-xs {display: none !important;}}@media (min-width: 30em) {.mantine-hidden-from-xs {display: none !important;}}@media (max-width: 47.99375em) {.mantine-visible-from-sm {display: none !important;}}@media (min-width: 48em) {.mantine-hidden-from-sm {display: none !important;}}@media (max-width: 63.99375em) {.mantine-visible-from-md {display: none !important;}}@media (min-width: 64em) {.mantine-hidden-from-md {display: none !important;}}@media (max-width: 74.99375em) {.mantine-visible-from-lg {display: none !important;}}@media (min-width: 75em) {.mantine-hidden-from-lg {display: none !important;}}@media (max-width: 89.99375em) {.mantine-visible-from-xl {display: none !important;}}@media (min-width: 90em) {.mantine-hidden-from-xl {display: none !important;}}@media (max-width: 23.99375em) {.mantine-visible-from-xxs {display: none !important;}}@media (min-width: 24em) {.mantine-hidden-from-xxs {display: none !important;}}
  </style>
  <div
    class="m_4451eb3a mantine-Center-root"
    style="margin-block: calc(6.25rem * var(--mantine-scale));"
  >
    <span
      class="m_4e3f22d7 m_5ae2e3c mantine-Loader-root"
      data-size="xl"
      style="--loader-size: var(--loader-size-xl);"
    >
      <span
        class="m_870c4af"
      />
      <span
        class="m_870c4af"
      />
      <span
        class="m_870c4af"
      />
    </span>
  </div>
</DocumentFragment>
`;
